hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -914,6 +911,30 @@ private TcpClientConnection onAuthenticated(TcpClientConnection connection,
         return connection;
     }
 
+    private void checkClientStateOnClusterIdChange(TcpClientConnection connection) {
+        if (activeConnections.isEmpty()) {
+            //We only have single connection established
+            if (failoverConfigProvided) {
+                //If failover is provided, and this single connection is established after failover logic kicks in
+                // (checked via `switchingToNextCluster`), then it is OK to continue. Otherwise, we force the failover logic
+                // to be used by throwing `ClientNotAllowedInClusterException`
+                if (switchingToNextCluster) {
+                    switchingToNextCluster = false;
+                } else {
```

Comment:
It is a behavior change, and it is actually a fix on the behavior. 

When this path is run there are two possibilities that happened in the cluster. 
1) There was a single cluster and it restarted. 
2) There were two clusters and the first one is closed. 

Note that there is no way for us to differentiate these two. In the second scenario, it could be the case that the same hostnames are used for both clusters and the user changed the hostname resolution before closing the first cluster. So even the addresses of the two clusters are the same for the client. 

There is no problem with the first case. And the behavior change for that looks unnecessary. But, it is an important fix for the second case. 
The problem of the second case is as follows. 
- Failover is configured. Now assume that two clusters have the same cluster name/credentials, meaning a client can switch between these two and will not notice any authentication failure. 
- First cluster has shutdown. 
- Also keep in mind that in failover config, users can provide different configurations for clusters.
- The client is will connect to the second cluster in the first attempt with the configuration of the first cluster. If we don't fail the connection, it will continue with the wrong configuration. We need to make sure that when failover config is provided, the client will remain connected to a cluster only if it is connected with the correct configuration.

Reply:
I see the problem with the second scenario. That makes sense then